### ISSUE TITLE
Convert 2D pixel point to 3D world space point

### ISSUE THREAD

**Issue by Upd4ting** [VulcanTechnologies/HoloLensCameraStream#37]
Hey I saw this method https://github.com/VulcanTechnologies/HoloLensCameraStream/blob/master/HoloLensVideoCaptureExample/Assets/CamStream/Scripts/LocatableCameraUtils.cs#L22
but I don't really know how to use it. That's returning a ray direction but then we need to enable the spatial mapping of the hololens in order to get a correct hit point. That will not be accurate using the spatial mapping cause the mesh generated isn't perfect. How can I do?

**Comment by Upd4ting** [comment]
Anyone here?

**Comment by SoylentGraham** [comment]
> How can I do?

Can you rephrase your question?

Yes, the spatial mapping isn't perfect.
Yes, that function gives you a world-space-ray.
Yes, people are here.

What exactly are you trying to achieve?

If it's not something specific to this library, (camera/hologram capture & camera intrinsics/extrinsics) your post might have more traction and response on http://www.stackoverflow.com

**Comment by SoylentGraham** [comment]
Try breaking down what you're trying to achieve into specific steps.

**Comment by Upd4ting** [comment]
I am tracking a color zone and then I need to convert the 2D location back to 3D location. I'm using this plugin to get the camera frame and I saw on the README here that was possible to get a 3D world coordinate from pixel coordinate. 
I posted also here: https://stackoverflow.com/questions/49075879/opencv-unity-get-3d-position-from-2d-image-position
but there I try to use more opencv related stuff to get 3D position.

**Comment by SoylentGraham** [comment]
That readme could be a little misleading, you can, (as you've pointed out with the function in the original post) that you can get a worldspace-ray from a 2D (screen space) position.

You would then use that ray, (as you say) to intersect with a plane, or geometry, or some kind of depth information.
The only depth information you get (as far as I'm aware) is the spatial mapping mesh.
So what you need to do, is take that ray, and find where it intersects the spatial mesh. This will give you your world space coordinate.

As you said, the spatial mesh is reduced, so it's not perfect.

The alternative to filling in that missing information is some system that will record the video and calculate depth information. This is more akin to photogrammetry, I'm not sure opencv provides any depth estimation functionality, and I certainly don't believe it would be fast enough to on mobile hardware. PLUS, you'd need a fast way to raycast into the data you'd get back (another thing that won't be immediately fast)
You could try capturing a few frames and treat them as stereo images and use various implementations to get a (very rough) depth map from stereo, but this again
- Won't be fast
- Will feel laggy as you need to use old frames and new frames
- Won't be perfect
- Won't work if the user doesn't move their head enough

You're probably going to have to use the spatial mesh, given the hardware limitations and hardware benefits.

If the hololens gave out it's depth map from the depth camera[s?], you'd be able to get a slightly better result, similar to what you see from a kinect.

**Comment by Upd4ting** [comment]
Thanks a lot for this detailled answer. Is there a way to update spatial mapping very fast?
